pkg/registry/core/service/strategy.go

```diff
@@ -205,6 +207,50 @@ func topologyKeysInUse(svc *api.Service) bool {
 	return len(svc.Spec.TopologyKeys) > 0
 }
 
+// This is an unusual call.  Service has a number of inter-linked fields and in
+// order to avoid breaking clients we try really hard to infer what users
+// meant.
+//
+// Services are effectively a discriminated union, where type is the
+// discriminator. Some fields just don't make sense with some types.
+func normalizeLinkedFields(newSvc *api.Service, oldSvc *api.Service) {
+	// For now we are only normalizing on updates.  This minimizes potential
+	// confusion since many of the cases we are handling here are pretty niche.
+	if oldSvc == nil {
+		return
+	}
+
+	// If the user is switching to a type that does not need a value (even
+	// "None" counts), we can wipe ClusterIP.  There are other fields that
+	// could also be wiped, but historically were ignored.  For compat, we
+	// leave it that way.
+	if needsClusterIP(oldSvc) && !needsClusterIP(newSvc) {
+		newSvc.Spec.ClusterIP = ""
+	}
+
+	// If the user is switching to a type that doesn't use NodePorts, we can
+	// wipe any allocated ports.
+	if needsNodePort(oldSvc) && !needsNodePort(newSvc) {
+		for i := range newSvc.Spec.Ports {
+			newSvc.Spec.Ports[i].NodePort = 0
+		}
+	}
+}
+
+func needsClusterIP(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeExternalName {
+		return false
+	}
+	return true
```

Comment:
I don't know how you feel about this in general:
```suggestion
	return svc.Spec.Type != api.ServiceTypeExternalName
```